### PR TITLE
Improve type printing (crashes)

### PR DESCRIPTION
**Pull request by soc**
Crashes now unroll more like this:

// SI-7406
class Arne[@specialized(Int) T](x: T) { lazy val y = x }

% scalac a.scala
error: assertion failed: Arne$mcI$sp.this.$asInstanceOf[Int.type]().unbox

  at scala.tools.nsc.ast.TreeGen.mkCast(TreeGen.scala:146)
  at scala.tools.nsc.Global$gen$.mkAttributedCast(Global.scala:106)
  at scala.tools.nsc.transform.Erasure$Eraser.cast(Erasure.scala:636)
  at scala.tools.nsc.transform.Erasure$Eraser.scala$tools$nsc$transform$Erasure$Eraser$$adaptToType(Erasure.scala:665)
  at scala.tools.nsc.transform.Erasure$Eraser.adapt(Erasure.scala:765)
  at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5345)
  at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedInternal(Typers.scala:5358)
  at scala.tools.nsc.typechecker.Typers$Typer.body$3(Typers.scala:5305)
  at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5310)
  at scala.tools.nsc.typechecker.Typers$Typer$$anonfun$60.apply(Typers.scala:3021)

```
 while compiling: a.scala
    during phase: mixin
 library version: version 2.11.0-20130608-020907-869ebd8bf2
compiler version: version 2.11.0-20130608-041104-feb660ccd7
```

one error found

### CODE REVIEW

**Comment by soc** [comment]
Review by @adriaanm. This is the third commit which I removed from https://github.com/scala/scala/pull/2746.

File: src/compiler/scala/tools/nsc/ast/Trees.scala [hunk 1]
```diff
@@ -14,6 +14,7 @@ import scala.reflect.internal.Flags.PARAMACCESSOR
 import scala.reflect.internal.Flags.PRESUPER
 import scala.reflect.internal.Flags.TRAIT
 import scala.compat.Platform.EOL
+import scala.reflect.internal.CompileRunAborted
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/Trees.scala:17]: Import added but not used.

**Comment by retronym** [comment]
DNLTGM yet. Calling `supplementErrorMessage` from `compileUnits` means that it will not report the tree currently being typechecked.

This commit does the commendable thing and tries to avoid falsely reporting "crashed while typing <sometree>" at some later phase that isn't actually typechecking said tree. But, with this in place, we have to record the current tree in the fatal error within `Typer#typed`.

Before this patch:

```
(remotes/origin/2784) /code/scala scalac-hash v2.11.0-M4 sandbox/crash-typer.scala
[info] v2.11.0-M4 => /Users/jason/usr/scala-v2.11.0-M4-0-g8b41240
error: trying to do lub/glb of typevar ?FT
error: scala.reflect.internal.FatalError:
     while compiling: sandbox/crash-typer.scala
        during phase: typer
     library version: version 2.11.0-20130710-171554-8b41240f3d
    compiler version: version 2.11.0-20130710-171554-8b41240f3d
  reconstructed args:

  last tree to typer: Ident($anon)
       tree position: line 4 of sandbox/crash-typer.scala
            tree tpe: Bug.H[Set]
              symbol: anonymous class $anon (flags: final)
   symbol definition: final class $anon extends Bug.H[Set] (a ClassSymbol)
      symbol package: <empty>
       symbol owners: anonymous class $anon -> value <local Bug> -> object Bug
           call site: package <root> in <none>

== Source file context for tree position ==

     1   trait H[F[_]]
     2   def f[F[_], T, FT <: F[T]](h : H[F]) = 1
     3   f(new H[Set]{})
     4 }
    at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:53)
    at scala.tools.nsc.Global.abort(Global.scala:248)
    at scala.reflect.internal.tpe.GlbLubs$class.stripType$1(GlbLubs.scala:212)
    at scala.reflect.internal.tpe.GlbLubs$$anonfun$20.apply(GlbLubs.scala:215)
    at scala.reflect.internal.tpe.GlbLubs$$anonfun$20.apply(GlbLubs.scala:215)
    at scala.collection.immutable.List.loop$1(List.scala:172)
    at scala.collection.immutable.List.mapConserve(List.scala:188)
    at scala.reflect.internal.tpe.GlbLubs$class.stripExistentialsAndTypeVars(GlbLubs.sca
       ....
```

With this patch:

```
(remotes/origin/2784) /code/scala qbin/scalac sandbox/crash-typer.scala
error: trying to do lub/glb of typevar ?FT

  at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:54)
  at scala.reflect.internal.tpe.GlbLubs$class.stripType$1(GlbLubs.scala:212)
  at scala.reflect.internal.tpe.GlbLubs$$anonfun$20.apply(GlbLubs.scala:215)
  at scala.reflect.internal.tpe.GlbLubs$$anonfun$20.apply(GlbLubs.scala:215)
  at scala.collection.immutable.List.loop$1(List.scala:172)
  at scala.collection.immutable.List.mapConserve(List.scala:188)
  at scala.reflect.internal.tpe.GlbLubs$class.stripExistentialsAndTypeVars(GlbLubs.scala:215)
  at scala.reflect.internal.tpe.GlbLubs$class.lub1$1(GlbLubs.scala:330)
  at scala.reflect.internal.tpe.GlbLubs$class.lub0$1(GlbLubs.scala:324)
  at scala.reflect.internal.tpe.GlbLubs$class.lub(GlbLubs.scala:419)

     while compiling: sandbox/crash-typer.scala
        during phase: typer
     library version: version 2.11.0-20130730-022954-ba51c90df3
    compiler version: version 2.11.0-20130730-022954-ba51c90df3

two errors found
```

**Comment by retronym** [comment]
Here's an approach that captures the tree of interest if an crash is detected in `Typer.typed` and stashes it in the current `Run`. https://github.com/retronym/scala/compare/topic/typer-crash-cleanup

**Comment by soc** [comment]
Should I close this and you create a new PR for this instead?

**Comment by retronym** [comment]
Yep

**Comment by soc** [comment]
Ok, thanks.

**Pull request closed without merging**